Projekt bazuje na informacjach udosptepnionych na stronie https://openweathermap.org/api przez firme **OpenWeather**.

Na potrzeby projektu ze wzgledu na przewidywane koszty zdecydowano nie laczyc sie bezposrednio z API strony tylko wybrano dane historyczne z okresu: ***01.01.1971 do 31.12.2023*** dla trzech przykladowych lokalizacji:
1. Bozanka, woj. pomorskie, gmina Szemud
2.
3.

### Parametry danych:

- **city_name** Nazwa miasta
- **lat** Geograficzne współrzędne lokalizacji (szerokość geograficzna)
- **lon** Geograficzne współrzędne lokalizacji (długość geograficzna)
- **main**
    - **main.temp** Temperatura [Celsjusz]
    - **main.temp_min** Minimalna temperatura w danym momencie. Jest to odchylenie od temperatury możliwej dla dużych miast i megalopolis geograficznie rozszerzonych (użyj tego parametru opcjonalnie) [Celsjusz].
    - **main.temp_max** Maksymalna temperatura w danym momencie. Jest to odchylenie od temperatury możliwej dla dużych miast i megalopolis geograficznie rozszerzonych (użyj tego parametru opcjonalnie) [Celsjusz].
    - **main.feels_like** Ten parametr temperatury uwzględnia ludzkie odczucie pogody [Celsjusz]
    - **main.pressure** Ciśnienie atmosferyczne (na poziomie morza), [hPa]
    - **main.humidity** Wilgotność, [%]
    - **main.dew_point** Temperatura atmosferyczna (zmieniająca się zgodnie z ciśnieniem i wilgotnością), poniżej której krople wody zaczynają kondensować się i tworzy się rosa. Jednostki – domyślnie: [Celsjusz]
- **wind**
    - **wind.speed** Prędkość wiatru. Jednostki – domyślnie: [metr/sek]
    - **wind.deg** Kierunek wiatru, [stopnie] (meteorologiczne)
    - **wind.gust** Poryw wiatru. Jednostki – domyślnie: [metr/sek]
- **clouds**
    - **clouds.all** Zachmurzenie, [%]
- **rain**
    - **rain.1h** Ilość opadów deszczu za ostatnią godzinę, [mm]
    - **rain.3h** Ilość opadów deszczu za ostatnie 3 godziny, [mm]
- **snow**
    - **snow.1h** Ilość opadów śniegu za ostatnią godzinę, [mm] (w stanie ciekłym)
    - **snow.3h** Ilość opadów śniegu za ostatnie 3 godziny, [mm] (w stanie ciekłym)
- **weather** (więcej informacji Kody stanu pogody) [-]
    - **weather.id** Identyfikator warunków pogodowych [-]
    - **weather.main** Grupa parametrów pogodowych (Deszcz, Śnieg, Ekstremalne itp.) [-]
    - **weather.description** Opis warunków pogodowych w grupie. Więcej informacji znajdziesz tutaj [-]
    - **weather.icon** Identyfikator ikony pogodowej [-]
- **visibility** Średnia widoczność, metry. Maksymalna wartość widoczności to 10km [m]
- **dt** Czas obliczenia danych, unix, UTC
- **dt_iso** Data i czas w formacie UTC 
- **timezone** Przesunięcie w sekundach względem UTC



In [10]:
import pandas as pd
import os

Ten skrypt najpierw tworzy listę wszystkich plików CSV w danym folderze, a następnie ładuje każdy z nich do osobnej ramki danych. Dodatkowo, przykład pokazuje, jak można dodać nazwę miasta do każdego zestawu danych na podstawie nazwy pliku. Po przetworzeniu wszystkich plików, ramki danych są łączone w jedną dużą ramkę, którą można następnie zapisać lub dalej przetwarzać.

Jeśli każdy plik CSV zawiera dane z innej lokalizacji i chcesz to uwzględnić, kod powyżej już dodaje nazwę miasta do każdego zbioru danych. Możesz dostosować ten proces do swoich potrzeb, na przykład przetwarzając nazwy plików w bardziej złożony sposób lub dodając inne metadane

In [11]:
# Ścieżka do folderu zawierającego pliki CSV
folder_path = '../2. Data/raw/'

# Lista plików CSV w folderze
csv_files = [file for file in os.listdir(folder_path) if file.endswith('.csv')]

# Lista do przechowywania danych z każdego pliku
all_data = []

for file in csv_files:
    file_path = os.path.join(folder_path, file)
    # Wczytanie danych z pliku
    df = pd.read_csv(file_path)
    
    # Możesz dodać tutaj dodatkową obróbkę danych dla każdego pliku, np. oczyszczanie danych
    # Przykład dodania nazwy miasta na podstawie nazwy pliku
    city_name = file.replace('.csv', '')  # Usunięcie rozszerzenia pliku z nazwy
    df['city_name'] = city_name  # Dodanie kolumny z nazwą miasta
    
    # Dodanie DataFrame do listy
    all_data.append(df)

# Połączenie wszystkich DataFrame w jedną
combined_data = pd.concat(all_data, ignore_index=True)

# Zapisanie połączonych danych do nowego pliku CSV
combined_data.to_csv('../2. Data/processed/combined_data.csv', index=False)

In [12]:
df_weather = pd.read_csv("../2. Data/processed/combined_data.csv")

In [13]:
df_weather.head(10)

dt                         dt_iso  timezone city_name        lat  \
0  283996800  1979-01-01 00:00:00 +0000 UTC      3600  Bozanska  54.473028   
1  284000400  1979-01-01 01:00:00 +0000 UTC      3600  Bozanska  54.473028   
2  284004000  1979-01-01 02:00:00 +0000 UTC      3600  Bozanska  54.473028   
3  284007600  1979-01-01 03:00:00 +0000 UTC      3600  Bozanska  54.473028   
4  284011200  1979-01-01 04:00:00 +0000 UTC      3600  Bozanska  54.473028   
5  284014800  1979-01-01 05:00:00 +0000 UTC      3600  Bozanska  54.473028   
6  284018400  1979-01-01 06:00:00 +0000 UTC      3600  Bozanska  54.473028   
7  284022000  1979-01-01 07:00:00 +0000 UTC      3600  Bozanska  54.473028   
8  284025600  1979-01-01 08:00:00 +0000 UTC      3600  Bozanska  54.473028   
9  284029200  1979-01-01 09:00:00 +0000 UTC      3600  Bozanska  54.473028   

         lon   temp  visibility  dew_point  feels_like  ...  wind_gust  \
0  18.302179 -15.81       600.0     -17.67      -22.81  ...        NaN   
1  18.302179 -15.84         NaN     -18.22      -22.84  ...        NaN   
2  18.302179 -15.97         NaN     -18.35      -22.97  ...        NaN   
3  18.302179 -16.29         NaN     -18.66      -23.29  ...        NaN   
4  18.302179 -16.04         NaN     -18.28      -23.04  ...        NaN   
5  18.302179 -15.90         NaN     -18.02      -22.90  ...        NaN   
6  18.302179 -15.92      1000.0     -17.78      -22.92  ...        NaN   
7  18.302179 -15.50         NaN     -17.50      -22.50  ...        NaN   
8  18.302179 -15.31         NaN     -17.31      -22.31  ...        NaN   
9  18.302179 -16.01         NaN     -18.52      -23.01  ...        NaN   

   rain_1h  rain_3h  snow_1h  snow_3h  clouds_all  weather_id  weather_main  \
0      NaN      NaN     0.26      NaN         100         601          Snow   
1      NaN      NaN     0.21      NaN         100         600          Snow   
2      NaN      NaN     0.19      NaN         100         600          Snow   
3      NaN      NaN     0.18      NaN         100         600          Snow   
4      NaN      NaN     0.17      NaN         100         600          Snow   
5      NaN      NaN     0.17      NaN         100         600          Snow   
6      NaN      NaN     0.18      NaN         100         601          Snow   
7      NaN      NaN     0.20      NaN         100         600          Snow   
8      NaN      NaN     0.21      NaN         100         600          Snow   
9      NaN      NaN     0.22      NaN         100         600          Snow   

   weather_description  weather_icon  
0                 snow           13n  
1           light snow           13n  
2           light snow           13n  
3           light snow           13n  
4           light snow           13n  
5           light snow           13n  
6                 snow           13n  
7           light snow           13n  
8           light snow           13d  
9           light snow           13d  

[10 rows x 28 columns]

In [14]:
df_weather.tail(10)

dt                         dt_iso  timezone city_name  \
406887  1704031200  2023-12-31 14:00:00 +0000 UTC      3600  Bozanska   
406888  1704034800  2023-12-31 15:00:00 +0000 UTC      3600  Bozanska   
406889  1704038400  2023-12-31 16:00:00 +0000 UTC      3600  Bozanska   
406890  1704042000  2023-12-31 17:00:00 +0000 UTC      3600  Bozanska   
406891  1704045600  2023-12-31 18:00:00 +0000 UTC      3600  Bozanska   
406892  1704049200  2023-12-31 19:00:00 +0000 UTC      3600  Bozanska   
406893  1704052800  2023-12-31 20:00:00 +0000 UTC      3600  Bozanska   
406894  1704056400  2023-12-31 21:00:00 +0000 UTC      3600  Bozanska   
406895  1704060000  2023-12-31 22:00:00 +0000 UTC      3600  Bozanska   
406896  1704063600  2023-12-31 23:00:00 +0000 UTC      3600  Bozanska   

              lat        lon  temp  visibility  dew_point  feels_like  ...  \
406887  54.473028  18.302179  2.60       650.0       1.58       -1.46  ...   
406888  54.473028  18.302179  2.36       150.0       1.49       -1.76  ...   
406889  54.473028  18.302179  1.95       150.0       1.24       -2.27  ...   
406890  54.473028  18.302179  2.11       100.0       1.39       -2.85  ...   
406891  54.473028  18.302179  2.11       200.0       1.39       -2.61  ...   
406892  54.473028  18.302179  1.86       300.0       1.15       -2.93  ...   
406893  54.473028  18.302179  1.71      1700.0       1.00       -3.12  ...   
406894  54.473028  18.302179  1.86      3600.0       1.00       -2.93  ...   
406895  54.473028  18.302179  2.11      4100.0       1.25       -2.61  ...   
406896  54.473028  18.302179  2.11      4600.0       1.25       -2.35  ...   

        wind_gust  rain_1h  rain_3h  snow_1h  snow_3h  clouds_all  weather_id  \
406887        NaN      NaN      NaN      NaN      NaN         100         741   
406888        NaN      NaN      NaN      NaN      NaN         100         741   
406889        NaN      NaN      NaN      NaN      NaN         100         741   
406890        NaN      NaN      NaN      NaN      NaN         100         741   
406891        NaN      NaN      NaN      NaN      NaN         100         741   
406892        NaN      NaN      NaN      NaN      NaN         100         741   
406893        NaN      NaN      NaN      NaN      NaN         100         701   
406894        NaN      NaN      NaN      NaN      NaN         100         701   
406895        NaN      NaN      NaN      NaN      NaN         100         701   
406896        NaN      NaN      NaN      NaN      NaN         100         701   

        weather_main  weather_description  weather_icon  
406887           Fog                  fog           50d  
406888           Fog                  fog           50n  
406889           Fog                  fog           50n  
406890           Fog                  fog           50n  
406891           Fog                  fog           50n  
406892           Fog                  fog           50n  
406893          Mist                 mist           50n  
406894          Mist                 mist           50n  
406895          Mist                 mist           50n  
406896          Mist                 mist           50n  

[10 rows x 28 columns]

In [15]:
#Spradzenie ogolnych informacji na temat danych.
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406897 entries, 0 to 406896
Data columns (total 28 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   dt                   406897 non-null  int64  
 1   dt_iso               406897 non-null  object 
 2   timezone             406897 non-null  int64  
 3   city_name            406897 non-null  object 
 4   lat                  406897 non-null  float64
 5   lon                  406897 non-null  float64
 6   temp                 406897 non-null  float64
 7   visibility           321941 non-null  float64
 8   dew_point            406897 non-null  float64
 9   feels_like           406897 non-null  float64
 10  temp_min             406897 non-null  float64
 11  temp_max             406897 non-null  float64
 12  pressure             406897 non-null  int64  
 13  sea_level            0 non-null       float64
 14  grnd_level           0 non-null       float64
 15  humidity         

In [16]:
missing_data = df_weather.isnull().sum()
missing_data

dt                          0
dt_iso                      0
timezone                    0
city_name                   0
lat                         0
lon                         0
temp                        0
visibility              84956
dew_point                   0
feels_like                  0
temp_min                    0
temp_max                    0
pressure                    0
sea_level              406897
grnd_level             406897
humidity                    0
wind_speed                  0
wind_deg                    0
wind_gust              380497
rain_1h                354089
rain_3h                406897
snow_1h                391325
snow_3h                406897
clouds_all                  0
weather_id                  0
weather_main                0
weather_description         0
weather_icon                0
dtype: int64

### Z przedstawionego zestawienia dla DataFrame df_weather wynika, że liczba brakujących danych w poszczególnych kolumnach jest następująca:

- **visibility**: Brakuje danych w 84,956 wierszach.
- **sea_level**: Wszystkie 406,897 danych są puste.
- **grnd_level**: Wszystkie 406,897 danych są puste.
- **wind_gust**: Brakuje danych w 380,497 wierszach.
- **rain_1h**: Brakuje danych w 354,089 wierszach.
- **rain_3h**: Wszystkie 406,897 danych są puste.
- **snow_1h**: Brakuje danych w 391,325 wierszach.
- **snow_3h**: Wszystkie 406,897 danych są puste.

Pozostałe kolumny mają pełne zestawienie danych bez braków.

Na tej podstawie można podjąć kilka działań w celu przygotowania zestawu danych do analizy:

- Kolumny **sea_level**, **grnd_level**, **rain_3h**, **snow_3h**: Ponieważ wszystkie dane w tych kolumnach są puste, najprawdopodobniej najlepszym rozwiązaniem będzie ich usunięcie, nie jest planowane ich wypełnienie danymi z innego źródła na tym etapie projektu.

- Kolumny z dużą ilością brakujących danych **wind_gust**, **rain_1h**, **snow_1h**: Jeżeli te dane są kluczowe dla analizy, można rozważyć ich uzupełnienie, na przykład przez interpolację, użycie średnich z innych kolumn, lub całkowite usunięcie wierszy, w których brak danych. Na tym etapie projektu zostala podjete deczyja o usunieciu tych kolumn.

- Kolumna **visibility**: Możliwe, że brakujące dane można uzupełnić, jeśli mają one wpływ na przeprowadzaną analizę, lub można te wiersze usunąć, jeżeli nie są one kluczowe. Na typ etapie proejktu zostala podjeta decyzja po uzupelniniu brakujacych danych. Do dyspozycji sa nastepujace metody:

**Imputacja średnią**:
Jeżeli brakujące dane są przypadkowe i nie mają wyraźnego wzorca, możliwym rozwiązaniem jest imputacja, czyli wypełnienie brakujących wartości średnią lub medianą z dostępnych danych dla visibility. W przypadku bardziej złożonych danych można zastosować modelowanie predykcyjne (na przykład regresję liniową) bazujące na innych zmiennych w zestawie danych, które mogą korelować z widocznością.

**Interpolacja**:
Jeśli dane są szeregami czasowymi, interpolacja może być dobrą metodą na oszacowanie brakujących wartości. Możesz użyć liniowej interpolacji, która wypełni brakujące wartości na podstawie wartości sąsiadujących w czasie, lub zastosować bardziej zaawansowane metody interpolacji, jak interpolacja splajnowa.

**Uzupełnienie na podstawie grupowania**:
Jeśli brakujące wartości visibility mogą być różne dla różnych warunków pogodowych lub innych zmiennych, możesz uzupełnić brakujące dane średnią lub medianą dla odpowiednich grup. Na przykład, jeśli widoczność jest zwykle niższa podczas opadów deszczu, możesz obliczyć średnią widoczność dla dni deszczowych i dni bez deszczu oddzielnie i użyć tych wartości do wypełnienia braków.

**Zastosowanie wartości z najbliższego sąsiedztwa (k-nearest neighbors - KNN)**:
Metoda ta polega na znajdowaniu 'k' najbliższych sąsiadów danej obserwacji z brakującymi danymi i wypełnieniu braku na podstawie wartości widoczności tych sąsiadów.

**Na potrzeby naszego proejktu uzyto metody imputacji sredniej.**

In [17]:
# Imputacja średnią dla kolumny 'visibility'
mean_visibility = df_weather['visibility'].mean()
df_weather['visibility'].fillna(mean_visibility, inplace=True)

In [18]:
# Usunięcie kolumn z pustymi danymi
columns_to_drop = ['sea_level', 'grnd_level', 'wind_gust', 'rain_1h', 'rain_3h', 'snow_1h', 'snow_3h']
df_weather.drop(columns=columns_to_drop, inplace=True)

In [19]:
missing_data = df_weather.isnull().sum()
missing_data

dt                     0
dt_iso                 0
timezone               0
city_name              0
lat                    0
lon                    0
temp                   0
visibility             0
dew_point              0
feels_like             0
temp_min               0
temp_max               0
pressure               0
humidity               0
wind_speed             0
wind_deg               0
clouds_all             0
weather_id             0
weather_main           0
weather_description    0
weather_icon           0
dtype: int64

In [20]:
df_weather.describe()

dt       timezone           lat           lon           temp  \
count  4.068970e+05  406897.000000  4.068970e+05  4.068970e+05  406897.000000   
mean   1.001183e+09    5565.283598  5.447303e+01  1.830218e+01       7.659756   
std    4.086583e+08    1792.397620  6.394892e-14  1.776359e-14       7.990468   
min    2.839968e+08    3600.000000  5.447303e+01  1.830218e+01     -27.640000   
25%    6.492024e+08    3600.000000  5.447303e+01  1.830218e+01       1.670000   
50%    1.006171e+09    7200.000000  5.447303e+01  1.830218e+01       7.440000   
75%    1.354464e+09    7200.000000  5.447303e+01  1.830218e+01      13.840000   
max    1.704064e+09    7200.000000  5.447303e+01  1.830218e+01      33.650000   

          visibility      dew_point     feels_like       temp_min  \
count  406897.000000  406897.000000  406897.000000  406897.000000   
mean     8250.477103       4.433560       5.322721       6.540989   
std      2655.097211       6.974908       9.714363       7.891047   
min         2.000000     -29.570000     -30.040000     -30.120000   
25%      8250.477103      -0.420000      -2.310000       0.860000   
50%      9999.000000       4.470000       4.920000       6.280000   
75%     10000.000000      10.020000      13.370000      12.770000   
max     10000.000000      25.830000      34.700000      32.620000   

            temp_max       pressure       humidity     wind_speed  \
count  406897.000000  406897.000000  406897.000000  406897.000000   
mean        8.615582    1014.273303      81.485194       4.401298   
std         8.051110      10.145607      14.524190       2.451244   
min       -21.250000     933.000000      12.000000       0.000000   
25%         2.350000    1008.000000      74.000000       2.640000   
50%         8.340000    1015.000000      86.000000       4.100000   
75%        14.800000    1021.000000      93.000000       6.000000   
max        36.660000    1050.000000     100.000000      34.000000   

            wind_deg     clouds_all     weather_id  
count  406897.000000  406897.000000  406897.000000  
mean      198.040610      48.325205     732.194219  
std        96.420851      38.407317     118.168751  
min         0.000000       0.000000     200.000000  
25%       130.000000       0.000000     701.000000  
50%       227.000000      60.000000     800.000000  
75%       270.000000      75.000000     803.000000  
max       360.000000     100.000000     804.000000

In [21]:
# Najpierw usunięcie "+0000 UTC" z końca stringów
df_weather['dt_iso'] = pd.to_datetime(df_weather['dt_iso'].str.replace(r' \+\d{4} UTC', '', regex=True))

In [22]:
df_weather.head()

dt              dt_iso  timezone city_name        lat        lon  \
0  283996800 1979-01-01 00:00:00      3600  Bozanska  54.473028  18.302179   
1  284000400 1979-01-01 01:00:00      3600  Bozanska  54.473028  18.302179   
2  284004000 1979-01-01 02:00:00      3600  Bozanska  54.473028  18.302179   
3  284007600 1979-01-01 03:00:00      3600  Bozanska  54.473028  18.302179   
4  284011200 1979-01-01 04:00:00      3600  Bozanska  54.473028  18.302179   

    temp   visibility  dew_point  feels_like  ...  temp_max  pressure  \
0 -15.81   600.000000     -17.67      -22.81  ...    -14.61       998   
1 -15.84  8250.477103     -18.22      -22.84  ...    -14.67       999   
2 -15.97  8250.477103     -18.35      -22.97  ...    -14.78       999   
3 -16.29  8250.477103     -18.66      -23.29  ...    -15.07      1000   
4 -16.04  8250.477103     -18.28      -23.04  ...    -14.79      1000   

   humidity  wind_speed  wind_deg  clouds_all  weather_id  weather_main  \
0        84        9.00        70         100         601          Snow   
1        80       10.66        83         100         600          Snow   
2        80       10.04        84         100         600          Snow   
3        80        9.58        83         100         600          Snow   
4        81        9.41        82         100         600          Snow   

  weather_description weather_icon  
0                snow          13n  
1          light snow          13n  
2          light snow          13n  
3          light snow          13n  
4          light snow          13n  

[5 rows x 21 columns]

In [23]:
# Następnie konwersja do datetime bez informacji o strefie czasowej
df_weather['dt_iso'] = pd.to_datetime(df_weather['dt_iso'], format='%Y-%m-%d %H:%M:%S', errors='coerce')

In [24]:
df_weather.head()

dt              dt_iso  timezone city_name        lat        lon  \
0  283996800 1979-01-01 00:00:00      3600  Bozanska  54.473028  18.302179   
1  284000400 1979-01-01 01:00:00      3600  Bozanska  54.473028  18.302179   
2  284004000 1979-01-01 02:00:00      3600  Bozanska  54.473028  18.302179   
3  284007600 1979-01-01 03:00:00      3600  Bozanska  54.473028  18.302179   
4  284011200 1979-01-01 04:00:00      3600  Bozanska  54.473028  18.302179   

    temp   visibility  dew_point  feels_like  ...  temp_max  pressure  \
0 -15.81   600.000000     -17.67      -22.81  ...    -14.61       998   
1 -15.84  8250.477103     -18.22      -22.84  ...    -14.67       999   
2 -15.97  8250.477103     -18.35      -22.97  ...    -14.78       999   
3 -16.29  8250.477103     -18.66      -23.29  ...    -15.07      1000   
4 -16.04  8250.477103     -18.28      -23.04  ...    -14.79      1000   

   humidity  wind_speed  wind_deg  clouds_all  weather_id  weather_main  \
0        84        9.00        70         100         601          Snow   
1        80       10.66        83         100         600          Snow   
2        80       10.04        84         100         600          Snow   
3        80        9.58        83         100         600          Snow   
4        81        9.41        82         100         600          Snow   

  weather_description weather_icon  
0                snow          13n  
1          light snow          13n  
2          light snow          13n  
3          light snow          13n  
4          light snow          13n  

[5 rows x 21 columns]

In [25]:
df_weather['city_name'] = df_weather['city_name'].replace('Custom location', 'Bozanska')

In [26]:
df_weather.head()

dt              dt_iso  timezone city_name        lat        lon  \
0  283996800 1979-01-01 00:00:00      3600  Bozanska  54.473028  18.302179   
1  284000400 1979-01-01 01:00:00      3600  Bozanska  54.473028  18.302179   
2  284004000 1979-01-01 02:00:00      3600  Bozanska  54.473028  18.302179   
3  284007600 1979-01-01 03:00:00      3600  Bozanska  54.473028  18.302179   
4  284011200 1979-01-01 04:00:00      3600  Bozanska  54.473028  18.302179   

    temp   visibility  dew_point  feels_like  ...  temp_max  pressure  \
0 -15.81   600.000000     -17.67      -22.81  ...    -14.61       998   
1 -15.84  8250.477103     -18.22      -22.84  ...    -14.67       999   
2 -15.97  8250.477103     -18.35      -22.97  ...    -14.78       999   
3 -16.29  8250.477103     -18.66      -23.29  ...    -15.07      1000   
4 -16.04  8250.477103     -18.28      -23.04  ...    -14.79      1000   

   humidity  wind_speed  wind_deg  clouds_all  weather_id  weather_main  \
0        84        9.00        70         100         601          Snow   
1        80       10.66        83         100         600          Snow   
2        80       10.04        84         100         600          Snow   
3        80        9.58        83         100         600          Snow   
4        81        9.41        82         100         600          Snow   

  weather_description weather_icon  
0                snow          13n  
1          light snow          13n  
2          light snow          13n  
3          light snow          13n  
4          light snow          13n  

[5 rows x 21 columns]

In [27]:
# Grupowanie danych po dniu (zignorowanie godziny)
df_weather['date'] = df_weather['dt_iso'].dt.date
daily_weather = df_weather.groupby('date').agg({
    # Agregacja danych pogodowych, np. średnia temperatura, suma opadów, maksymalne zachmurzenie itp.
    'city_name': 'first',
    'lat': 'first',
    'lon': 'first',
    'temp': 'mean',
    'feels_like': 'mean',
    'temp_min': 'min',
    'temp_max': 'max',
    'pressure': 'mean',
    'humidity': 'mean',
    'wind_speed': 'mean',
    #'rain_1h': 'sum',
    #'rain_3h': 'sum',
    #'snow_1h': 'sum',
    #'snow_3h': 'sum',
    'clouds_all': 'max',
    'weather_main': lambda x: x.mode()[0] if not x.mode().empty else 'Unknown'  # Najczęstsze zjawisko pogodowe
}).reset_index()

In [28]:
df_weather.head(10)

dt              dt_iso  timezone city_name        lat        lon  \
0  283996800 1979-01-01 00:00:00      3600  Bozanska  54.473028  18.302179   
1  284000400 1979-01-01 01:00:00      3600  Bozanska  54.473028  18.302179   
2  284004000 1979-01-01 02:00:00      3600  Bozanska  54.473028  18.302179   
3  284007600 1979-01-01 03:00:00      3600  Bozanska  54.473028  18.302179   
4  284011200 1979-01-01 04:00:00      3600  Bozanska  54.473028  18.302179   
5  284014800 1979-01-01 05:00:00      3600  Bozanska  54.473028  18.302179   
6  284018400 1979-01-01 06:00:00      3600  Bozanska  54.473028  18.302179   
7  284022000 1979-01-01 07:00:00      3600  Bozanska  54.473028  18.302179   
8  284025600 1979-01-01 08:00:00      3600  Bozanska  54.473028  18.302179   
9  284029200 1979-01-01 09:00:00      3600  Bozanska  54.473028  18.302179   

    temp   visibility  dew_point  feels_like  ...  pressure  humidity  \
0 -15.81   600.000000     -17.67      -22.81  ...       998        84   
1 -15.84  8250.477103     -18.22      -22.84  ...       999        80   
2 -15.97  8250.477103     -18.35      -22.97  ...       999        80   
3 -16.29  8250.477103     -18.66      -23.29  ...      1000        80   
4 -16.04  8250.477103     -18.28      -23.04  ...      1000        81   
5 -15.90  8250.477103     -18.02      -22.90  ...      1000        82   
6 -15.92  1000.000000     -17.78      -22.92  ...      1000        84   
7 -15.50  8250.477103     -17.50      -22.50  ...      1001        83   
8 -15.31  8250.477103     -17.31      -22.31  ...      1002        83   
9 -16.01  8250.477103     -18.52      -23.01  ...      1002        79   

   wind_speed  wind_deg  clouds_all  weather_id  weather_main  \
0        9.00        70         100         601          Snow   
1       10.66        83         100         600          Snow   
2       10.04        84         100         600          Snow   
3        9.58        83         100         600          Snow   
4        9.41        82         100         600          Snow   
5        9.34        81         100         600          Snow   
6        5.00       130         100         601          Snow   
7        8.42        79         100         600          Snow   
8        8.02        77         100         600          Snow   
9        7.67        75         100         600          Snow   

   weather_description weather_icon        date  
0                 snow          13n  1979-01-01  
1           light snow          13n  1979-01-01  
2           light snow          13n  1979-01-01  
3           light snow          13n  1979-01-01  
4           light snow          13n  1979-01-01  
5           light snow          13n  1979-01-01  
6                 snow          13n  1979-01-01  
7           light snow          13n  1979-01-01  
8           light snow          13d  1979-01-01  
9           light snow          13d  1979-01-01  

[10 rows x 22 columns]

In [29]:
daily_weather.head(11)

date city_name        lat        lon       temp  feels_like  \
0   1979-01-01  Bozanska  54.473028  18.302179 -14.100000  -20.617500   
1   1979-01-02  Bozanska  54.473028  18.302179  -9.037917  -15.179167   
2   1979-01-03  Bozanska  54.473028  18.302179 -10.780000  -16.679167   
3   1979-01-04  Bozanska  54.473028  18.302179 -15.627083  -20.063750   
4   1979-01-05  Bozanska  54.473028  18.302179  -5.102083  -11.947500   
5   1979-01-06  Bozanska  54.473028  18.302179  -6.948333  -13.033750   
6   1979-01-07  Bozanska  54.473028  18.302179  -7.670833  -14.644583   
7   1979-01-08  Bozanska  54.473028  18.302179   0.062500   -5.922083   
8   1979-01-09  Bozanska  54.473028  18.302179  -0.335417   -5.145417   
9   1979-01-10  Bozanska  54.473028  18.302179  -1.374167   -7.143750   
10  1979-01-11  Bozanska  54.473028  18.302179  -3.272083   -9.855417   

    temp_min  temp_max     pressure   humidity  wind_speed  clouds_all  \
0     -19.14     -9.46  1002.416667  82.000000    6.157917         100   
1     -14.14     -7.40  1004.791667  85.375000    4.449583         100   
2     -16.33     -6.84  1004.500000  87.041667    3.805833         100   
3     -21.14     -9.08  1008.916667  86.083333    2.346250         100   
4     -11.22     -2.42  1014.500000  84.750000    6.470417         100   
5     -13.14     -2.08  1027.666667  87.291667    4.378750          93   
6     -13.43     -1.46  1024.958333  82.250000    9.538333         100   
7      -3.14      0.92  1014.500000  95.583333    7.611667         100   
8      -1.49      0.69  1009.541667  95.458333    4.762500         100   
9      -4.29      0.16   999.291667  91.708333    6.078750         100   
10     -5.24     -1.08   992.500000  90.500000    6.949167         100   

   weather_main  
0          Snow  
1        Clouds  
2        Clouds  
3        Clouds  
4          Snow  
5        Clouds  
6        Clouds  
7        Clouds  
8        Clouds  
9        Clouds  
10       Clouds

Zapisanie ramki do pliku po przetworzeniu:

In [30]:
daily_weather.to_csv('../2. Data/processed/combined_data_processed.csv')